src/libraries/Common/tests/Tests/System/TimeProviderTests.cs

```diff
@@ -0,0 +1,380 @@
+                                        case 2:
+                                            s.Period = 400;
+                                            s.Timer.Change(TimeSpan.FromMilliseconds(s.Period), TimeSpan.FromMilliseconds(s.Period));
+                                            break;
+
+                                        case 4:
+                                            mres.Set();
+                                            s.Timer.Dispose();
+                                            break;
+                                    }
+
+                                    s.UtcNow = DateTimeOffset.UtcNow;
+                                },
+                            state,
+                            TimeSpan.FromMilliseconds(state.Period), TimeSpan.FromMilliseconds(state.Period));
+
+            mres.Wait(TimeSpan.FromMilliseconds(30000));
+
+            Assert.Equal(4, state.Counter);
+            Assert.Equal(400, state.Period);
+            Assert.True(MaxMilliseconds >= state.TotalTicks / TimeSpan.TicksPerMillisecond, $"The total fired periods {state.TotalTicks / TimeSpan.TicksPerMillisecond}ms expected not exceeding the expected max {MaxMilliseconds}");
+        }
+
+        [Fact]
+        public void FastClockTest()
+        {
+            FastClock fastClock = new FastClock();
+
+            for (int i = 0; i < 20; i++)
+            {
+                DateTimeOffset fastNow = fastClock.UtcNow;
+                DateTimeOffset now = DateTimeOffset.UtcNow;
+
+                Assert.True(fastNow > now, $"Expected {fastNow} > {now}");
+
+                fastNow = fastClock.LocalNow;
+                now = DateTimeOffset.Now;
+
+                Assert.True(fastNow > now, $"Expected {fastNow} > {now}");
+            }
+
+            Assert.Equal(TimeSpan.TicksPerSecond, fastClock.TimestampFrequency);
+
+            long stamp1 = fastClock.GetTimestamp();
+            long stamp2 = fastClock.GetTimestamp();
+
+            Assert.Equal(stamp2 - stamp1, fastClock.GetElapsedTime(stamp1, stamp2).Ticks);
+        }
+
+        public static IEnumerable<object[]> TimersProvidersListData()
+        {
+            yield return new object[] { TimeProvider.System };
+            yield return new object[] { new FastClock() };
+        }
+
+        [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]
+        [MemberData(nameof(TimersProvidersListData))]
+        public static void CancellationTokenSourceWithTimer(TimeProvider provider)
+        {
+            //
+            // Test out some int-based timeout logic
+            //
+            CancellationTokenSource cts = new CancellationTokenSource(Timeout.InfiniteTimeSpan, provider); // should be an infinite timeout
+            CancellationToken token = cts.Token;
+            ManualResetEventSlim mres = new ManualResetEventSlim(false);
+            CancellationTokenRegistration ctr = token.Register(() => mres.Set());
```

Comment:
Rather than creating an MRES and registering with the token, for test simplicity you could just use the token's WaitHandle and WaitOne on it.  Or you could `await Task.Delay(-1, token)`, which will await until cancellation is requested.